### PR TITLE
Adding support for Context Parallelism using Deepseed's DistributedAt…

### PR DESCRIPTION
**Pull request by bhargaveede**
Adding support for Context Parallelism using Deepseed's DistributedAttention

This PR adds support to enable Context Parallelism for Llama models using deepspeed.
This feature can be enabled by using context_parallel_size flag.

This feature enables us to train/eval longer context lengths by parallelizing inputs across Context Parallel group.
For Attention, It uses Deepspeed's DistributedAttention which gathers the sequences for all heads and distributes the heads across the Context Parallel Group so that Attention per head has entire context and are distributed within the Context Parallel group.
Once the attention is done, Outputs are scattered based on sequence length across the group.

Verified Llama3.1 8B and Llama 3.1 70B finetuning for 32K seq length on 8 ranks using this feature.

- Could fit 32K sequence length Llama 3.1-8B LoRa finetuning on 4 ranks with selective recompute instead of gradient_checkpointing.
- Could fit 32k sequence length for Llama 3.1-70B LoRa finetuning on 4 ranks and 8 ranks with gradient_checkpointing 

Llama 3.1-8B command
HL_DS_DISTRIBUTED_ATTENTION_SEQ_DIM=1 MASTER_ADDR=127.0.0.1 MASTER_PORT=12345   python3 ./optimum-habana-fork/examples/gaudi_spawn.py  --world_size 8 --use_deepspeed ./optimum-habana-fork/examples/language-modeling/run_lora_clm.py  --dataset_name tatsu-lab/alpaca --bf16 True --output_dir /tmp/lora_out --max_seq_len 32768 --per_device_train_batch_size 1 --per_device_eval_batch_size 1 --save_strategy no --learning_rate 0.0004 --warmup_ratio 0.03 --lr_scheduler_type "constant" --logging_steps 1 --dataset_concatenation --do_train --use_habana --throughput_warmup_steps 3 --lora_rank 8 --lora_target_modules "q_proj" "v_proj" "k_proj" "o_proj" --attn_softmax_bf16 True --validation_split_percentage 4 --flash_attention_causal_mask True --evaluation_strategy epoch --pipelining_fwd_bwd --use_lazy_mode --use_flash_attention True --deepspeed ./optimum-habana-fork/examples/language-modeling/llama3_ds_zero1_config.json --num_train_epochs 3 --eval_delay 3 --do_eval --lora_alpha 16 --lora_dropout 0.05 --gradient_accumulation_steps 4 --flash_attention_recompute True --context_parallel_size 4 --model_name_or_path meta-llama/Llama-3.1-8B

Test added to verify the context parallelism:
https://github.com/huggingface/optimum-habana/pull/1501/files#diff-
0741a50beca4b08d354933485499f735f9b5493841e8f3af0e89b16ae1e04af4R978

Note:

DistributedAttention iIndices](https://github.com/huggingface/optimum-habana/pull/1501/files#diff-30aeee6868dd1de34878aca0583f57bb5b0dd9a2a8511a80e9a6b2645f39ce6bR490) are initialized as scatter_idx 1 and gather_idx 2 as for Llama Query States has shape [B,N,S,H] - [Batch Size, Num Heads, Seq Length, Head Dim]
As we want to gather on sequence length [Dim 2 of B,N,S,H] and scatter heads [Dim 1 of B,N,S,H]
Other models who want to Integrate DistributedAttention have to adjust the indices based on the shape.

# What does this PR do?

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

**Comment by bhargaveede** [comment]
@regisss @libinta 
Please review the changes which add support for Context Parallelism.
This is needed for 1.19 release

File: optimum/habana/accelerate/accelerator.py [hunk 1]
```diff
@@ -59,6 +59,8 @@
 from accelerate.utils.other import is_compiled_module
 from torch.optim.lr_scheduler import LRScheduler
 
+from optimum.habana import parallel_state
```
Comment (regisss) [inline optimum/habana/accelerate/accelerator.py:62]:
```suggestion
from .. import parallel_state
```
Let's use relative imports to stay consistent with the rest of the library

File: optimum/habana/accelerate/data_loader.py [hunk 2]
```diff
@@ -22,6 +22,8 @@
 )
 from torch.utils.data import BatchSampler, DataLoader, IterableDataset
 
+from optimum.habana import parallel_state
```
Comment (regisss) [inline optimum/habana/accelerate/data_loader.py:25]:
```suggestion
from .. import parallel_state
```

File: optimum/habana/accelerate/state.py [hunk 3]
```diff
@@ -19,6 +19,7 @@
 from accelerate.state import AcceleratorState, PartialState
 from accelerate.utils import is_deepspeed_available, parse_choice_from_env, parse_flag_from_env
 
+from optimum.habana import parallel_state
```
Comment (regisss) [inline optimum/habana/accelerate/state.py:22]:
```suggestion
from .. import parallel_state
```

File: optimum/habana/distributed/contextparallel.py [hunk 4]
```diff
@@ -0,0 +1,30 @@
+import torch
+
+from optimum.habana.parallel_state import (
```
Comment (regisss) [inline optimum/habana/distributed/contextparallel.py:3]:
```suggestion
from ..parallel_state import (
```

File: optimum/habana/parallel_state.py
Comment (regisss) [inline optimum/habana/parallel_state.py:1]: Is this file inspired from another repo? If yes, can you link to the source?

File: optimum/habana/transformers/models/llama/modeling_llama.py [hunk 5]
```diff
@@ -21,6 +21,8 @@
 )
 from transformers.utils import is_torchdynamo_compiling
 
+from optimum.habana import parallel_state
```
Comment (regisss) [inline optimum/habana/transformers/models/llama/modeling_llama.py:24]:
```suggestion
from .... import parallel_state
```

File: optimum/habana/transformers/models/llama/modeling_llama.py [hunk 6]
```diff
@@ -1398,7 +1434,19 @@ def forward(
...
             # Enable model parallelism
             shift_labels = shift_labels.to(shift_logits.device)
-            loss = loss_fct(shift_logits, shift_labels)
+            # Collect losses from context parallel group
+            # Each rank in group calculates loss on partial outputs
+            if (
+                parallel_state.sequence_parallel_is_initialized()
+                and parallel_state.get_sequence_parallel_world_size() > 1
+            ):
+                from optimum.habana.distributed.contextparallel import _get_loss_from_context_parallel
```
Comment (regisss) [inline optimum/habana/transformers/models/llama/modeling_llama.py:1443]:
```suggestion
                from ....distributed.contextparallel import _get_loss_from_context_parallel
```

File: optimum/habana/parallel_state.py
Comment (regisss) [inline optimum/habana/parallel_state.py:1]: Why not putting this file in `optimum/habana.distributed`?

File: optimum/habana/transformers/training_args.py [hunk 7]
```diff
@@ -190,6 +190,11 @@ class GaudiTrainingArguments(TrainingArguments):
         },
     )
 
+    context_parallel_size: Optional[int] = field(
+        default=1,
+        metadata={"help": ("Determines how many ranks are divided into context parallel group.")},
+    )
```
Comment (regisss) [inline optimum/habana/transformers/training_args.py:196]: Should we mention this new arg in the documentation to let people know how to use it? Or is the support rather experimental so far?

File: tests/test_examples.py [hunk 8]
```diff
@@ -234,6 +234,7 @@ def to_test(
             "codellama/CodeLlama-13b-Instruct-hf",
             "MIT/ast-finetuned-speech-commands-v2",
             "meta-llama/LlamaGuard-7b",
+            "huggyllama/llama-7b",
```
Comment (regisss) [inline tests/test_examples.py:237]: Can we perform the test with a more recent version of Llama? This one is Llama v1.

**Review by regisss — COMMENTED** [review]
This cannot be merged before SynapseAI v1.19 is released right?

**Review by yeonsily — COMMENTED** [review]
Can you please also run all of the llama CI tests to make sure this doesn't affect to the current number?

File: optimum/habana/accelerate/accelerator.py (hunk 1 above)
Comment (bhargaveede) [reply optimum/habana/accelerate/accelerator.py:62]: Sure

File: optimum/habana/parallel_state.py
Comment (bhargaveede) [reply optimum/habana/parallel_state.py:1]: Yes. I will link the source

File: optimum/habana/transformers/training_args.py (hunk 7 above)
Comment (bhargaveede) [reply optimum/habana/transformers/training_args.py:196]: We can add this to documentation.
However, I am not sure about the issues related to DistributedAttention esp. for models other than Llama.

File: optimum/habana/parallel_state.py
Comment (bhargaveede) [reply optimum/habana/parallel_state.py:1]: @regisss 
Source File:
https://github.com/NVIDIA/Megatron-LM/blob/67a50f2f9b2f12b1361ac0f21fc3410b23e07d5e/megatron/core/parallel_state.py#L4

LICENSE for the smae:
https://github.com/NVIDIA/Megatron-LM?tab=License-1-ov-file#readme

File: tests/test_examples.py (hunk 8 above)
Comment (bhargaveede) [reply tests/test_examples.py:237]: We can add the test on any version of Llama. 
We added v1 as for llama2 and 3 we need hugging face token authorization to access the model. If not, it will fail.
Is there a way to pass the token? (or) what do you suggest?

File: optimum/habana/accelerate/state.py [hunk 9]
```diff
@@ -85,6 +86,9 @@ def __init__(self, cpu: bool = False, **kwargs):
                 if self.device is None:
                     # TODO: replace by `torch.device("hpu", self.local_process_index)` when hpu:x is supported
                     self.device = torch.device("hpu")
+                if not is_deepspeed_available():
+                    context_parallel_size = 1
+                parallel_state.initialize_model_parallel(sequence_parallel_size=context_parallel_size, use_fp8=False)
```
[the remark this answers is not in the document]
Comment (bhargaveede) [reply optimum/habana/accelerate/state.py:91]: We didn't test fp8. There might be some additional things needed for fp8, though.
Only difference between fp8 and bf16 is that AMAX_REDUCTION_GROUPS are created. There might be some additional handling for calculation of amax values with CP.

File: optimum/habana/transformers/training_args.py (hunk 7 above)
Comment (regisss) [reply optimum/habana/transformers/training_args.py:196]: Okay, let's keep it that way for now then :+1:

File: tests/test_examples.py (hunk 8 above)
Comment (regisss) [reply tests/test_examples.py:237]: For instance I use this script for the DeepSpeed CI: https://github.com/huggingface/optimum-habana/blob/main/tests/ci/slow_tests_deepspeed.sh
It takes a token as an argument to log in. But not sure you do it like that.

File: optimum/habana/parallel_state.py
Comment (bhargaveede) [reply optimum/habana/parallel_state.py:1]: We can put it.
Can I do the restructuring later in a separate commit?

File: tests/test_examples.py (hunk 8 above)
Comment (bhargaveede) [reply tests/test_examples.py:237]: We added this test to check the working of CP feature.
For llama 2 and 3.1, We can add one more for long sequence length.
Can I add that later along with restructuring?

Comment (regisss) [reply tests/test_examples.py:244]: Yes sure, but I would like to do that before release

Comment (bhargaveede) [reply tests/test_examples.py:244]: @regisss Do we have any existing tests for Llama2 or Llama3.1 as part of OH. 
How do we pass token there? Can you point out the same?

Comment (bhargaveede) [reply tests/test_examples.py:244]: If not, I think we should add token similar to https://github.com/huggingface/optimum-habana/blob/main/tests/ci/slow_tests_deepspeed.sh for https://github.com/huggingface/optimum-habana/blob/main/tests/ci/slow_tests_8x.sh as well

I will update the test to Llama3.1. Can you add token to the slow_tests_8x.sh?
Will add it along with restructuring in separate commit.

If you agree for that. Can you get this merged?

**Comment by regisss** [comment]
> @regisss style check is failing in a different file (not part of this PR). Can you check?

Yeah that was happening because a PR was merged yesterday without passing the style check. I rebased your branch so everything should be fine now.

File: tests/test_examples.py (hunk 8 above)
Comment (regisss) [reply tests/test_examples.py:244]: Okay, let's do that :+1:

**Review by regisss — APPROVED** [review]

File: tests/test_examples.py (hunk 8 above)
Comment (regisss) [reply tests/test_examples.py:244]: @bhargaveede I just pushed a commit to add HF login to the 8x slow tests: https://github.com/huggingface/optimum-habana/commit/9f9b41e6b718f9972b8c349368bb88fa0584bd4c
